x-pack/metricbeat/module/gcp/metadata_cache.go

```diff
@@ -0,0 +1,247 @@
+type CacheRegistry struct {
+	logger    *logp.Logger
+	dataMutex sync.RWMutex
+
+	compute  *Cache[*computepb.Instance]
+	cloudsql *Cache[*sqladmin.DatabaseInstance]
+	redis    *Cache[*redispb.Instance]
+	dataproc *Cache[*dataproc.Cluster]
+}
+
+// NewCacheRegistry creates a new cache registry.
+func NewCacheRegistry(logger *logp.Logger, refreshInterval time.Duration) *CacheRegistry {
+	return &CacheRegistry{
+		logger:   logger,
+		compute:  NewCache[*computepb.Instance](refreshInterval),
+		cloudsql: NewCache[*sqladmin.DatabaseInstance](refreshInterval),
+		redis:    NewCache[*redispb.Instance](refreshInterval),
+		dataproc: NewCache[*dataproc.Cluster](refreshInterval),
+	}
+}
+
+// updateCache is a generic helper to update cache
+func updateCache[T any](cache *Cache[T], update map[string]T) {
+	if cache.data == nil {
+		cache.data = make(map[string]T)
+	}
+	for k, v := range update {
+		cache.data[k] = v
+	}
+	cache.meta.LastRefreshed = time.Now()
+}
+
+// ensureCacheFresh is a generic helper to ensure cache freshness
+func ensureCacheFresh[T any](
+	r *CacheRegistry,
+	cache *Cache[T],
+	ctx context.Context,
+	cacheName string,
+	fetchFunc func(context.Context) (map[string]T, error),
+) error {
+	r.dataMutex.RLock()
+	expired := cache.meta.IsExpired()
+	r.dataMutex.RUnlock()
+
+	if !expired {
+		r.logger.Debugf("%s cache is fresh.", cacheName)
+		return nil
+	}
+
+	r.logger.Debugf("%s cache potentially expired, acquiring fetch lock...", cacheName)
+	cache.fetchMutex.Lock()
+	defer cache.fetchMutex.Unlock()
+	r.logger.Debugf("%s fetch lock acquired.", cacheName)
+
+	r.dataMutex.RLock()
+	expired = cache.meta.IsExpired()
+	r.dataMutex.RUnlock()
+
+	if !expired {
+		r.logger.Debugf("%s cache was refreshed by another goroutine while waiting for fetch lock.", cacheName)
+		return nil
+	}
+
+	r.logger.Infof("%s cache expired, executing fetch function...", cacheName)
+
+	expBackoff := backoff.NewExponentialBackOff()
+	expBackoff.InitialInterval = 1 * time.Second
+	expBackoff.MaxInterval = 30 * time.Second
+	expBackoff.MaxElapsedTime = 2 * time.Minute
+
+	backoffWithRetry := backoff.WithMaxRetries(expBackoff, 3)
+
+	var newData map[string]T
+	operation := func() error {
+		var err error
+		newData, err = fetchFunc(ctx)
+		if err != nil {
+			r.logger.Warnf("%s cache fetch attempt failed: %v", cacheName, err)
+			return err
+		}
+		return nil
+	}
+
+	err := backoff.Retry(operation, backoff.WithContext(backoffWithRetry, ctx))
+	if err != nil {
+		r.logger.Errorf("%s cache fetch function failed after retries: %v", cacheName, err)
+		return fmt.Errorf("%s cache fetch failed: %w", cacheName, err)
+	}
+
+	r.dataMutex.Lock()
+	updateCache(cache, newData)
+	r.dataMutex.Unlock()
+
+	r.logger.Infof("%s cache successfully refreshed.", cacheName)
+	return nil
+}
+
+// UpdateComputeCache appends the provided map to the Compute instance cache.
+func (r *CacheRegistry) UpdateComputeCache(update map[string]*computepb.Instance) {
```

Comment:
nit : regarding updates, seems they are not used. Is there any plan to use them or do you see the need ?

Reply:
initially I thought I will use them in each metadata collector but now that is being done in `ensureCacheFresh`, I will remove them